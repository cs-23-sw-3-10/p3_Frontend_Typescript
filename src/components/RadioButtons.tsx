import React from "react";
import "./RadioButtons.css";

//Declaring types of props
type RadioButtonsProps= {
    options: string[];
    name: string;
    defaultCheckedOption: string;
}

/**
 * @param {string[]}  options: Array of options represented by the radio buttons
 * @param {string}  name: Name of the group of radio buttons. Only one button in the group can be checked at given time.
 * @param {string}  defaultCheckedOption: option to be checked as default. To-do: We should throw error if defaultOptionChecked does not correspond to one of the options
 * @returns
 */
const RadioButtons= ({options,name,defaultCheckedOption}:RadioButtonsProps) => {
    const handleOptionChange = (changeEvent: React.ChangeEvent<HTMLInputElement>) => {
        console.log("Option selected: " + changeEvent.target.value);//Here we should make call to server. 
    };

    return (
        <div onChange={handleOptionChange} className="radiobuttons">
            {options.map((option) => (
                <label key={option}> {/*good practice to give key as unique identifier of elements in the list.  */}
                    <input
                        type="radio"
                        value={option}
                        name={name}
                        defaultChecked={option === defaultCheckedOption}
                    />
                    {option}
                </label>
            ))}
        </div>
    );
};

export default RadioButtons;
