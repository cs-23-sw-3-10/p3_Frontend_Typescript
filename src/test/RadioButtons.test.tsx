import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect"; //Extends Jest matchers witch matchers provided by the jest-dom library. the code uses the matcher "toBeChecked".
import RadioButtons from "../components/RadioButtons";

describe("RadioButtons Component", () => {
    //Snapshot test for correct rendering
    it("renders correctly", () => { 
        const component = renderer.create( //generates virtual represnetation of the component
            <RadioButtons
                options={["Edit", "View"]}
                name={"editOrView"}
                defaultCheckedOption="View"
            ></RadioButtons>
        );
        const tree = component.toJSON();//generates snapshot 
        expect(tree).toMatchSnapshot();//compares the snapshot with a previously stored snapshot
    });

    //Change test using render, fireevent and screen. 
    it("changes correctly", () => {
        const options = ["Option 1", "Option 2", "Option 3"];
        const defaultCheckedOption = "Option 1";

        render( // render component in div-container appended to document.body
            <RadioButtons
                options={options}
                name="radioGroup"
                defaultCheckedOption={defaultCheckedOption}
            />
        );

        // Get radio buttons by their labels. screen has every type of query used to find elements on the page (what page exactly?), including getBtLabelText
        const option1 = screen.getByLabelText("Option 1");
        const option2 = screen.getByLabelText("Option 2");
        const option3 = screen.getByLabelText("Option 3");

        // Check the initial state
        expect(option1).toBeChecked();
        expect(option2).not.toBeChecked();
        expect(option3).not.toBeChecked();

        // Simulate clicking Option 2
        fireEvent.click(option2);

        //Check the updated state
        expect(option1).not.toBeChecked();
        expect(option2).toBeChecked();
        expect(option3).not.toBeChecked();
    });
});
