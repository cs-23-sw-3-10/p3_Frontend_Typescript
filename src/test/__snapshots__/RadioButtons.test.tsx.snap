// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtons Component renders correctly 1`] = `
<div
  className="radiobuttons"
  onChange={[Function]}
>
  <label>
     
    <input
      defaultChecked={false}
      name="editOrView"
      type="radio"
      value="Edit"
    />
    Edit
  </label>
  <label>
     
    <input
      defaultChecked={true}
      name="editOrView"
      type="radio"
      value="View"
    />
    View
  </label>
</div>
`;
